### PR TITLE
makefiles: Generate proper dependency files when using ccache

### LINKED ISSUE #12807: make: rebuild doesn't apply to header files
#### Description
When rebuilding with a changed value in a header this changed value is not taken into account.

#### Steps to reproduce the issue
I created [a very simple test application](https://gist.github.com/miri64/f5a47100b363fe6b6178652480b659a2). When building and running for `native` in current master (d0d6e53), I first get as expected:

```
$ make clean all term
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change'
Building application "tests_" for "native" with MCU "native".

"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/core
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native/periph
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/auto_init
   text	   data	    bss	    dec	    hex	filename
  20251	    568	  47716	  68535	  10bb7	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change/bin/native/tests_.elf
/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change/bin/native/tests_.elf  
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-1051-gd0d6e)
0205
```

If I then change `GNRC_NETAPI_MSG_TYPE_ACK` *in the header* (I incremented it by one to `0x0206`) and then rebuild *without clean* I still get the output `0205`:

```
make: Entering directory '/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change'
Building application "tests_header_change" for "native" with MCU "native".

"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/boards/native/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/core
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native/periph
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/cpu/native/vfs
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/drivers/periph_common
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys
"make" -C /home/mlenders/Repositories/RIOT-OS/RIOT/sys/auto_init
   text	   data	    bss	    dec	    hex	filename
  23303	    568	  47716	  71587	  117a3	/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change/bin/native/tests_header_change.elf
/home/mlenders/Repositories/RIOT-OS/RIOT/tests/header_change/bin/native/tests_header_change.elf  
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

main(): This is RIOT! (Version: 2020.01-devel-1051-gd0d6e)
0205
```

I confirmed that the same happens on `samr21-xpro` and also with RIOT versions as old as 2017.01.

#### Expected results
After the second build, the app should show `0206` (if `GNRC_NETAPI_MSG_TYPE_ACK` was incremented e.g.).

#### Actual results
After the second build, the app still shows `0205`.

#### Versions
```

Operating System Environment
-----------------------------
       Operating System: "Ubuntu" "19.10 (Eoan Ermine)"
                 Kernel: Linux 5.3.0-23-generic x86_64 x86_64

Installed compiler toolchains
-----------------------------
             native gcc: gcc (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
      arm-none-eabi-gcc: arm-none-eabi-gcc (GNU Tools for Arm Embedded Processors 8-2018-q4-major) 8.2.1 20181213 (release) [gcc-8-branch revision 267074]
                avr-gcc: avr-gcc (GCC) 5.4.0
       mips-mti-elf-gcc: missing
             msp430-gcc: msp430-gcc (GCC) 4.6.3 20120301 (mspgcc LTS 20120406 unpatched)
   riscv-none-embed-gcc: missing
   xtensa-esp32-elf-gcc: missing
   xtensa-lx106-elf-gcc: missing
                  clang: clang version 9.0.0-2 (tags/RELEASE_900/final)

Installed compiler libs
-----------------------
   arm-none-eabi-newlib: "3.0.0"
    mips-mti-elf-newlib: missing
riscv-none-embed-newlib: missing
xtensa-esp32-elf-newlib: missing
xtensa-lx106-elf-newlib: missing
               avr-libc: "2.0.0" ("20150208")

Installed development tools
---------------------------
                  cmake: cmake version 3.13.4
               cppcheck: Cppcheck 1.88
                doxygen: 1.8.13
                    git: git version 2.20.1
                   make: GNU Make 4.2.1
                openocd: Open On-Chip Debugger 0.10.0+dev-00910-g4dbcb1e7 (2019-06-17-16:24)
                 python: Python 2.7.17rc1
                python2: Python 2.7.17rc1
                python3: Python 3.7.5rc1
                 flake8: 3.7.8 (mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.7.5rc1 on Linux
             coccinelle: missing
```

I checked this with RIOT versions d0d6e53ff4, e4fb634e3f (2018.01), and 158463e696 (2017.01).

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
In #12807 it has been reported that when a change is made in a header file (which is a dependency of a given object file) make does not take this into account, in consequence it does not re-trigger a build.

The cause of this issue are the dependency files generated by the compiler (placed under `$(BINDIR)/$(MODULE)/<filename>.d`). When ccache is used, if the `CCACHE_BASEDIR` environmental variable is set, then it will turn all paths into relative (see [ccache docs](https://ccache.dev/manual/latest.html)). This has the advantage of allowing one to reuse the cache across multiple RIOT folders, but in our case it passes a relative path to the object in the call to gcc. 

Here you can see a snippet of ccache log, the actual call to gcc. Note that paths are relative:
<details><summary><b>ccache log on master</b></summary>

```
 Running preprocessor
[2019-11-28T19:50:38.589228 23100] Executing /usr/bin/gcc -Werror -Wall -Wextra -pedantic
-std=gnu11 -m32 -fstack-protector-all -ffunction-sections -fdata-sections -fno-common -Wall
-Wextra -Wmissing-include-dirs -fno-delete-null-pointer-checks -fdiagnostics-color
-Wstrict-prototypes -Wold-style-definition -gz -Wformat=2 -Wformat-overflow -Wformat-truncation
-DRIOT_FILE_RELATIVE="examples/hello-world/main.c" -DRIOT_FILE_NOPATH="main.c"
-DDEVELHELP -DDEBUG_ASSERT_VERBOSE -DRIOT_APPLICATION="hello-world"
-DBOARD_NATIVE="native" -DRIOT_BOARD=BOARD_NATIVE -DCPU_NATIVE="native"
-DRIOT_CPU=CPU_NATIVE -DMCU_NATIVE="native"
-DRIOT_MCU=MCU_NATIVE -include bin/native/riotbuild/riotbuild.h -I../../core/include
-I../../drivers/include -I../../sys/include -I../../boards/native/include -DNATIVE_INCLUDES
-I../../boards/native/include -I../../core/include -I../../drivers/include -I../../cpu/native/include
-I../../sys/include -I../../cpu/native/include -MD -MP -MF bin/native/application_hello-world/main.d
-MQ bin/native/application_hello-world/main.o -E main.c
```
</details>

The compiler then will generate rules with relative paths:
<details><summary><b>main.d file with relative path to object</b></summary>

```makefile
bin/native/application_hello-world/main.o: main.c \
 /usr/include/stdc-predef.h bin/native/riotbuild/riotbuild.h \
 /usr/include/stdio.h /usr/include/bits/libc-header-start.h \
 /usr/include/features.h /usr/include/sys/cdefs.h \
 /usr/include/bits/wordsize.h /usr/include/bits/long-double.h \
 /usr/include/gnu/stubs.h /usr/include/gnu/stubs-32.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stddef.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdarg.h \
 /usr/include/bits/types.h /usr/include/bits/timesize.h \
 /usr/include/bits/typesizes.h /usr/include/bits/time64.h \
 /usr/include/bits/types/__fpos_t.h /usr/include/bits/types/__mbstate_t.h \
 /usr/include/bits/types/__fpos64_t.h /usr/include/bits/types/__FILE.h \
 /usr/include/bits/types/FILE.h /usr/include/bits/types/struct_FILE.h \
 /usr/include/bits/stdio_lim.h /usr/include/bits/sys_errlist.h \
 ../../sys/include/net/gnrc/netapi.h ../../core/include/thread.h \
 ../../core/include/clist.h ../../core/include/list.h \
 ../../core/include/cib.h ../../core/include/assert.h \
 ../../core/include/panic.h ../../core/include/kernel_defines.h \
 ../../core/include/msg.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdint.h /usr/include/stdint.h \
 /usr/include/bits/wchar.h /usr/include/bits/stdint-intn.h \
 /usr/include/bits/stdint-uintn.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdbool.h \
 ../../core/include/kernel_types.h /usr/include/inttypes.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed/limits.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed/syslimits.h \
 /usr/include/limits.h /usr/include/bits/posix1_lim.h \
 /usr/include/bits/local_lim.h /usr/include/linux/limits.h \
 /usr/include/bits/posix2_lim.h ../../cpu/native/include/cpu_conf.h \
 ../../core/include/sched.h ../../core/include/bitarithm.h \
 ../../core/include/native_sched.h ../../sys/include/net/netopt.h \
 ../../sys/include/net/gnrc/nettype.h ../../sys/include/net/ethertype.h \
 ../../sys/include/net/protnum.h ../../sys/include/net/gnrc/pkt.h \
 /usr/include/stdlib.h /usr/include/bits/waitflags.h \
 /usr/include/bits/waitstatus.h /usr/include/bits/floatn.h \
 /usr/include/bits/floatn-common.h /usr/include/sys/types.h \
 /usr/include/bits/types/clock_t.h /usr/include/bits/types/clockid_t.h \
 /usr/include/bits/types/time_t.h /usr/include/bits/types/timer_t.h \
 /usr/include/endian.h /usr/include/bits/endian.h \
 /usr/include/bits/byteswap.h /usr/include/bits/uintn-identity.h \
 /usr/include/sys/select.h /usr/include/bits/select.h \
 /usr/include/bits/types/sigset_t.h /usr/include/bits/types/__sigset_t.h \
 /usr/include/bits/types/struct_timeval.h \
 /usr/include/bits/types/struct_timespec.h \
 /usr/include/bits/pthreadtypes.h /usr/include/bits/thread-shared-types.h \
 /usr/include/bits/pthreadtypes-arch.h /usr/include/alloca.h \
 /usr/include/bits/stdlib-float.h ../../core/include/kernel_types.h
```
</details>

The dependency files are included then in `Makefile.base`, but as the rules do not match out pattern in the objects rule, the objects are not updated when a dependency (e.g. a header file) is changed.

To solve this problem, we need absolute paths in the targets of the *.d files. While using ccache 3.6 I managed this just by not setting `CCACHE_BASEDIR`, so paths would be absolute. @cgundogan reported that this does not seem to fix the issue for version 3.7.6. That is why I modified the [gcc flags](https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html) to pass an explicit absolute path as the target of the generated rule (only doing this for ccache does not solve the issue, as it detects the path and relativizes it). This should not affect users who don't use ccache.

With this PR, dependency files have absolute paths, and any changes to the files the objects depend on will be taken into account.

<details><summary><b>main.d with this PR</b></summary>

```makefile
/home/leandro/Work/RIOT/examples/hello-world/bin/native/application_hello-world/main.o: \
 /home/leandro/Work/RIOT/examples/hello-world/main.c \
 /usr/include/stdc-predef.h \
 /home/leandro/Work/RIOT/examples/hello-world/bin/native/riotbuild/riotbuild.h \
 /usr/include/stdio.h /usr/include/bits/libc-header-start.h \
 /usr/include/features.h /usr/include/sys/cdefs.h \
 /usr/include/bits/wordsize.h /usr/include/bits/long-double.h \
 /usr/include/gnu/stubs.h /usr/include/gnu/stubs-32.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stddef.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdarg.h \
 /usr/include/bits/types.h /usr/include/bits/timesize.h \
 /usr/include/bits/typesizes.h /usr/include/bits/time64.h \
 /usr/include/bits/types/__fpos_t.h /usr/include/bits/types/__mbstate_t.h \
 /usr/include/bits/types/__fpos64_t.h /usr/include/bits/types/__FILE.h \
 /usr/include/bits/types/FILE.h /usr/include/bits/types/struct_FILE.h \
 /usr/include/bits/stdio_lim.h /usr/include/bits/sys_errlist.h \
 /home/leandro/Work/RIOT/sys/include/net/gnrc/netapi.h \
 /home/leandro/Work/RIOT/core/include/thread.h \
 /home/leandro/Work/RIOT/core/include/clist.h \
 /home/leandro/Work/RIOT/core/include/list.h \
 /home/leandro/Work/RIOT/core/include/cib.h \
 /home/leandro/Work/RIOT/core/include/assert.h \
 /home/leandro/Work/RIOT/core/include/panic.h \
 /home/leandro/Work/RIOT/core/include/kernel_defines.h \
 /home/leandro/Work/RIOT/core/include/msg.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdint.h /usr/include/stdint.h \
 /usr/include/bits/wchar.h /usr/include/bits/stdint-intn.h \
 /usr/include/bits/stdint-uintn.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include/stdbool.h \
 /home/leandro/Work/RIOT/core/include/kernel_types.h \
 /usr/include/inttypes.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed/limits.h \
 /usr/lib/gcc/x86_64-linux-gnu/8/include-fixed/syslimits.h \
 /usr/include/limits.h /usr/include/bits/posix1_lim.h \
 /usr/include/bits/local_lim.h /usr/include/linux/limits.h \
 /usr/include/bits/posix2_lim.h \
 /home/leandro/Work/RIOT/cpu/native/include/cpu_conf.h \
 /home/leandro/Work/RIOT/core/include/sched.h \
 /home/leandro/Work/RIOT/core/include/bitarithm.h \
 /home/leandro/Work/RIOT/core/include/native_sched.h \
 /home/leandro/Work/RIOT/sys/include/net/netopt.h \
 /home/leandro/Work/RIOT/sys/include/net/gnrc/nettype.h \
 /home/leandro/Work/RIOT/sys/include/net/ethertype.h \
 /home/leandro/Work/RIOT/sys/include/net/protnum.h \
 /home/leandro/Work/RIOT/sys/include/net/gnrc/pkt.h /usr/include/stdlib.h \
 /usr/include/bits/waitflags.h /usr/include/bits/waitstatus.h \
 /usr/include/bits/floatn.h /usr/include/bits/floatn-common.h \
 /usr/include/sys/types.h /usr/include/bits/types/clock_t.h \
 /usr/include/bits/types/clockid_t.h /usr/include/bits/types/time_t.h \
 /usr/include/bits/types/timer_t.h /usr/include/endian.h \
 /usr/include/bits/endian.h /usr/include/bits/byteswap.h \
 /usr/include/bits/uintn-identity.h /usr/include/sys/select.h \
 /usr/include/bits/select.h /usr/include/bits/types/sigset_t.h \
 /usr/include/bits/types/__sigset_t.h \
 /usr/include/bits/types/struct_timeval.h \
 /usr/include/bits/types/struct_timespec.h \
 /usr/include/bits/pthreadtypes.h /usr/include/bits/thread-shared-types.h \
 /usr/include/bits/pthreadtypes-arch.h /usr/include/alloca.h \
 /usr/include/bits/stdlib-float.h \
 /home/leandro/Work/RIOT/core/include/kernel_types.h
```
</details>

Not setting `CCACHE_BASEDIR` will have the side effect of not sharing the cache between different RIOT folders that the user may have. If the user really wants this (and does not care about having to clean for every build), the variable can always be set externally.

### Testing procedure
Run the test under these two conditions:
1. **Using ccache**
- Make sure that you either have ccache in your path with symbolic links to it (as a wrapper of the compiler) or `CCACHE` environmental variable set: this depends on your installation.
2. **Not using ccache**
- Set the environmental variable `CCACHE_DISABLE`

*TIP*: To check if ccache is being used you can run `ccache -s` to see the statistics, the last access will change every time it's used.

**Test**:
Using the example application provided in #12807 (or a similar example):
- Run `rm -Rf bin` in the application directory.
-  Build `make all term`. You should see the current value in of the header file.
- Check that the `main.d` file has an absolute path to the object.
- Change the header file updating the value, and run `make all term` (NO CLEAN). `main.o` should be the only object re-built, and the number should be updated.

### Issues/PRs references
Fixes #12807

### CODE REVIEW

File: Makefile.base [hunk 1]
```diff
@@ -100,25 +100,25 @@ $(OBJC): $(BINDIR)/$(MODULE)/%.o: %.c $(RIOTBUILD_CONFIG_HEADER_C)
 	$(Q)$(CCACHE) $(CC) \
 		-DRIOT_FILE_RELATIVE=\"$(patsubst $(RIOTBASE)/%,%,$(abspath $<))\" \
 		-DRIOT_FILE_NOPATH=\"$(notdir $<)\" \
-		$(CFLAGS) $(INCLUDES) -MD -MP -c -o $@ $(abspath $<)
+		$(CFLAGS) $(INCLUDES) -MQ $@ -MD -MP -c -o $@ $(abspath $<)
```
Comment (cgundogan) [inline Makefile.base]: Not sure if this is necessary, but the docs [1] put apostrophes around the parameter:
```
 -MQ '$(objpfx)foo.o'
```
That might help for cases where `$@` has whitespaces (though currently this can never happen, I suppose).

https://gcc.gnu.org/onlinedocs/gcc/Preprocessor-Options.html

**Comment by cgundogan** [comment]
I confirm that this fixes #12807 for my setup (ccache 3.7.6)

**Comment by kaspar030** [comment]
> While using ccache 3.6 I managed this just by not setting `CCACHE_BASEDIR`, so paths would be absolute.

this probably breaks ci caching...

File: Makefile.include [hunk 2]
```diff
@@ -55,15 +54,13 @@ __DIRECTORY_VARIABLES := \
...
   BINDIR \
-  CCACHE_BASEDIR \
   GITCACHE \
   PKGDIRBASE \
   DLCACHE_DIR \
   #
 
 # Make all paths absolute.
 override RIOTBASE       := $(abspath $(RIOTBASE))
-override CCACHE_BASEDIR := $(abspath $(CCACHE_BASEDIR))
```
Comment (kaspar030) [inline Makefile.include]: this is essential for CI, unsetting this multiplies CI cache usage, causing 50-100% slowdown.

Comment (cgundogan) [reply Makefile.include]: Isn't the correct `CCACHE_BASEDIR` for murdock set here?
https://github.com/RIOT-OS/RIOT/blob/master/.murdock#L234

**Comment by MrKevinWeiss** [comment]
> this is essential for CI, unsetting this multiplies CI cache usage, causing 50-100% slowdown.

Ok lets see.  If does increase build time we will need either add something in RIOT (maybe a `RIOT_CI_BUILD` type thing) or something for murdock to do so.

File: Makefile.include (hunk 2 above)
Comment (kaspar030) [reply Makefile.include]: Ah, that's correct, murdock shouldn't be affected.

**Comment by kaspar030** [comment]
> Ah, that's correct, murdock shouldn't be affected.

This will affect local caching when building in multiple directories, but I guess that's preferable to broken depencencies.

**Comment by kaspar030** [comment]
>  If does increase build time

build time was 31min (+50%), I'll trigger a rebuild.

**Comment by kaspar030** [comment]
> build time was 31min (+50%)

That could actually be caused by the new "-MQ" option (which would cause cache misses), not by messing up caching. A couple of rebuilds would show.

**Comment by cgundogan** [comment]
27 minutes says murdock. Let's trigger again.

**Comment by kaspar030** [comment]
> 27 minutes says murdock. Let's trigger again.

I have a script that creates per-branch ccache stats for murdock, which I used for debugging the "-g" issue before. I'll run it for this PR (and current master) once the current run went through.

**Comment by kaspar030** [comment]
29min. Will get some cache stats now.

**Comment by kaspar030** [comment]
Here are the results: [master](https://gist.github.com/938f4a1739fca64c56c09a82196ba114), [this pr](https://gist.github.com/a33e253e56c3c004ddc9f41e3905d1f7).

TL;DR no difference, the PR doesn't seem to affect ccache on CI negatively.

**Comment by leandrolanzieri** [comment]
> Here are the results: [master](https://gist.github.com/938f4a1739fca64c56c09a82196ba114), [this pr](https://gist.github.com/a33e253e56c3c004ddc9f41e3905d1f7).
> 
> TL;DR no difference, the PR doesn't seem to affect ccache on CI negatively.

@kaspar030 thanks a lot for this! It was really helpful.

@cgundogan I added the apostrophes.

**Comment by leandrolanzieri** [comment]
Thanks everyone for testing and reviewing!
